### PR TITLE
[STACK-2180][STACK-2167][STACK-2168]: Dynamic Interface Detection(Attachement and Detachment) for Active-Standby topology

### PR DESCRIPTION
**Pull request by NehaKembalkarA10**
## Description
Added support for dynamic interface attachement detection in case of loadbalancers and members creation and for dynamic interface detachment detection in case of loadbalancers and members deletion for ACTIVE_STANDBY topology.
Also fixed existing code issues which were causing some failures in case of ACTIVE_STANDBY topology.

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-2167
https://a10networks.atlassian.net/browse/STACK-2168
https://a10networks.atlassian.net/browse/STACK-2180

## Technical Approach
- As `CreateHealthMonitorOnVThunder` task in `get_new_lb_networking_subflow` was getting called on each loadbalancer creation in ACTIVE_STANDBY topology, it was throwing an error as `"object specified already exists"`. So executing that task only on 1st loadbalancer creation now to fix that issue.
- Executing `get_vrrp_subflow` only on 1st load_balancer creation as it's not required on 2nd onwards loadbalancer creation as vrrp gets configured only once.
- Modified task `GetComputeForProject` by getting vthunder from `get_vthunder_by_project_id_and_role()` for getting compute_id of vThunder specific to its role stored in database.
- Added support for dynamic interface attachment by adding task `AmphoraePostVIPPlug` in `get_new_lb_networking_subflow` for ACTIVE_STANDBY topology in case of loadbalancer creation
- Created a task `GetVThunderInterface` for getting interface on vMaster and vBlade which is supposed to be enabled after attachment.
- Created a task `EnableInterfaceForActiveStandbyVThunder` for enabling interfaces on vMaster and vBlade.
- If interface got from `GetVThunderInterface` is of vMaster i.e **ifnum_master** then **setInterface()** will be called for enabling that interafce and if interface got from `GetVThunderInterface` is of vBlade i.e **ifnum_backup** then **device-context.switch(2, None)** should get called first from vMaster and then **setInterface()** should get called to enable the vBlade's interface.
- Added the tasks `GetVThunderInterface` and `EnableInterfaceForActiveStandbyVThunder` in `get_new_lb_networking_subflow`  after `AmphoraePostVIPPlug` sothat the attached interfaces of vMaster and vBlade will get enabled.
- Using the task `EnableInterface` only for STANDALONE topology by adding a topology check there.
- Added support for dynamic interface attachment during member creation for ACTIVE_STANDBY topology.
- Added support for dynamic interface detachment by adding task `AmphoraePostNetworkUnplug` to `delete_LB_flow` and `delete_member_flow` for ACTIVE_STANDBY topology in case of loadbalancer and member deletion respectively.

## Test Cases
- Given a vThunder, when VCS is not enabled on it, then interface should get enabled there without checking role of it.
- Given a vThunder, when VCS is enabled on it and its role at that moment is vMaster, then ifnum_master should be returned.
- Given a vThunder, when VCS is enabled on it and its role at that moment is vBlade, then ifnum_backup should be returned.
- Given a vThunder, if it is vMaster with having ifnum_master not None, then setInterface() should get called to enable that interface.
- Given a vThunder, if it is of STANDALONE topology, then EnableInterface task should get executed.
- Given a vThunder, if it is of ACTIVE_STANDBY topology, then EnableInterface task should not get executed.

## Manual Testing
**Case1:**
Create lb1 with subnet as vip_subnet1
    -2 vThunders will get booted, one is vMaster and another is vBlade
    -Reload/Reboot will not get performed
    -AXAPI call will be executed for saving acos_version in db
    -interface ethernet 1/1 and interface ethernet 2/1 will get configured on vMaster and vBlade.

Create lb2 with subnet as vip_subnet1
    -Reload/Reboot will not get performed
    -No AXAPI call for saving acos_version in db, Instead acos_version will be fetched from existing vthunder entry created for
      lb1 and saved in db.
    -No new interface will get attached for lb2

Create lb3 with subnet as vip_subnet2
    -Reload/Reboot will get performed depending on acos version
    -No AXAPI call for saving acos_version in db, Instead acos_version will be fetched from existing vthunder entry created for
     lb1 and saved in db.
    -interface ethernet 1/2 and interface ethernet 2/2 will get configured for lb4 on vMaster and vBlade.

**Case2:**
Create lb1 with subnet as vip_subnet1
    -2 vThunders will get booted, one is vMaster and another is vBlade
    -Reload/Reboot will not get performed
    -AXAPI call will be executed for saving acos_version in db
    -interface ethernet 1/1 and interface ethernet 2/1 will get configured on vMaster and vBlade.

Create a member with subnet as vip_subnet2
    -Reload/Reboot will get performed depending on acos_version
    -interface ethernet 1/2 and interface ethernet 2/2 will get configured on vMaster and vBlade.

**Case3:**
Create lb1 with subnet as vip_subnet1
    -2 vThunders will get booted, one is vMaster and another is vBlade
    -Reload/Reboot will not get performed
    -AXAPI call will be executed for saving acos_version in db
    -interface ethernet 1/1 and interface ethernet 2/1 will get configured on vMaster and vBlade.

Create lb2 with subnet as vip_subnet1
    -Reload/Reboot will not get performed
    -No AXAPI call for saving acos_version in db, Instead acos_version will be fetched from existing vthunder entry created for
      lb1 and saved in db.
    -No new interface will get attached for lb2

Create lb3 with subnet as vip_subnet2
    -Reload/Reboot will get performed depending on acos version
    -No AXAPI call for saving acos_version in db, Instead acos_version will be fetched from existing vthunder entry created for
     lb1 and saved in db.
    -interface ethernet 1/2 and interface ethernet 2/2 will get configured for lb3 on vMaster and vBlade.

Delete lb3
    -Reload/Reboot will get performed
    -interface ethernet 1/2 and interface ethernet 2/2 will get detached from vMaster and vBlade.

Delete lb2
    -No reload/reboot
    -No interface detachment

Delete lb1
    -vMaster and vBlade will get deleted

**Case4:**
Create lb1 with subnet as vip_subnet1
    -2 vThunders will get booted, one is vMaster and another is vBlade
    -Reload/Reboot will not get performed
    -AXAPI call will be executed for saving acos_version in db
    -interface ethernet 1/1 and interface ethernet 2/1 will get configured on vMaster and vBlade.

Create a member with subnet as vip_subnet2
    -Reload/Reboot will get performed depending on acos_version
    -interface ethernet 1/2 and interface ethernet 2/2 will get configured on vMaster and vBlade.

Delete member
    -Reload/Reboot will get performed depending on acos_version
    -interface ethernet 1/2 and interface ethernet 2/2 will get detached from vMaster and vBlade.

### CODE REVIEW

**Review by hthompson-a10 — COMMENTED** [review]
I have some code level comments if we decide to stick with this solution, but I don't think all of this is necessary for the Amphora side of things. Furthermore, the current flow is unlikely to be failover friendly and may have unintended side effects.

We also make the assumption that this is the only disable device in the list which is okay, but it could lead to hard to debug issues especially after failover events.

There could also be a situation where another NIC has gone down for some external reason. In that case, we could end up enabling that one instead of our most recently added one.

The other option is to always enable all of the disabled interfaces, but this may have unintended side effects later on. If interfaces have gone down, we'd want to report that. We could run into the case where an interface has gone down, and right after we create a new load balancer so we attempt to bring it back up. This could result in failures that we don't expect.

I propose the following as an alternative:

Knows and Assumptions:
- When an interface is attached, it must be the same on the Master and Backup Amphora. Unlike hardware devices, we control the creation of these interface from the beginning. So each should have consistent values.
- The current number of ethernet interfaces on the device
- The interface that was added last will be in the disabled state
- The current device id's state/role (Master or Backup)

Unknowns:
- The current device id's relation to role
- The current state of the cluster (the cluster could be in partial failed state)

Limitations & Concerns:
- I think we can ignore the cluster state consideration for now until we've addressed the failover logic.


Therefore we could instead:
1. Get the vThunder
2. Get the current NIC count (Let this equal NIC_COUNT)
3. Attach the new Interface
4. Assert that iface w/ NIC #(NIC_COUNT + 1) exists `and` is disabled
5. Enable the interface for NIC #(NIC_COUNT + 1)
6. If Active-Standby topology:
    1. get the Backup vThunder
    2. Assert that iface w / NIC #(NIC_COUNT + 1) exists in the device-context of the backup `and`  is disabled
    3. Enable the interface for NIC #(NIC_COUNT + 1)

We may need to do this in an `EnableAmphoraInterface` task and rename `EnableInterface` to `EnableHardwareInterface`.

**Review by hthompson-a10 — CHANGES_REQUESTED** [review]
Forgot to change over to "Request Changes". Doing that here.

File: a10_octavia/controller/worker/tasks/vthunder_tasks.py [hunk 1]
```diff
@@ -943,3 +945,66 @@ def execute(self, added_ports, loadbalancer, vthunder):
...
+                            LOG.debug(
+                                "Fetched the ethernet interface for Backup vThunder: %s", vthunder.id)
+                vthunder.ip_address = vcs_summary['vcs-summary']['oper']['member-list'][0]['ip-list'][0]['ip']
+            return vthunder, ifnum_master, ifnum_backup
+        except (acos_errors.ACOSException, req_exceptions.ConnectionError) as e:
+            LOG.exception("Failed to fetch ethernet interface Backup vThunder: %s", str(e))
+            raise e
+
+
+class EnableInterfaceForActiveStandbyVThunder(VThunderBaseTask):
```
[the remark this answers is not in the document]
Comment (NehaKembalkarA10) [reply a10_octavia/controller/worker/tasks/vthunder_tasks.py]: Removed the `EnableInterfaceForActiveStandbyVThunder` and added all logic in `EnableInterface` itself.

File: a10_octavia/controller/worker/tasks/vthunder_tasks.py [hunk 2]
```diff
@@ -943,3 +945,66 @@ def execute(self, added_ports, loadbalancer, vthunder):
...
+
+            vcs_summary = self.axapi_client.system.action.get_vcs_summary_oper()
+            vcs_enabled = vcs_summary['vcs-summary']['oper']['vcs-enabled']
+            interfaces = self.axapi_client.interface.get_list()
+
+            if vcs_enabled == "Invalid":
+                for i in range(len(interfaces['interface']['ethernet-list'])):
+                    if interfaces['interface']['ethernet-list'][i]['action'] == "disable":
+                        ifnum = interfaces['interface']['ethernet-list'][i]['ifnum']
+                        self.axapi_client.system.action.setInterface(ifnum)
```
[the remark this answers is not in the document]
Comment (NehaKembalkarA10) [reply a10_octavia/controller/worker/tasks/vthunder_tasks.py]: Removed this logic from here and handled this case in flow

File: a10_octavia/controller/worker/flows/a10_load_balancer_flows.py [hunk 3]
```diff
@@ -290,24 +312,51 @@ def get_new_lb_networking_subflow(self, topology):
...
             new_LB_net_subflow.add(
                 a10_database_tasks.GetBackupVThunderByLoadBalancer(
                     name=a10constants.GET_BACKUP_VTHUNDER_BY_LB,
                     requires=constants.LOADBALANCER,
                     provides=a10constants.BACKUP_VTHUNDER))
+            new_LB_net_subflow.add(vthunder_tasks.UpdateAcosVersionInVthunderEntry(
+                name=a10constants.UPDATE_ACOS_VERSION_FOR_BACKUP_VTHUNDER,
+                requires=constants.LOADBALANCER,
+                rebind={a10constants.VTHUNDER: a10constants.BACKUP_VTHUNDER}))
+            new_LB_net_subflow.add(a10_database_tasks.GetBackupVThunderByLoadBalancer(
```
[the remark this answers is not in the document]
Comment (NehaKembalkarA10) [reply a10_octavia/controller/worker/flows/a10_load_balancer_flows.py:335]: This is here for getting updated Backup vThunder entry after updating acos_version in vthunders entry for Backup

[the remark this answers is not in the document]
Comment (hthompson-a10) [reply a10_octavia/controller/worker/flows/a10_load_balancer_flows.py:339]: Why not return the the vThunder model with the update acos_version in the "UpdateAcosVersionInVthunderEntry" task?

[the remark this answers is not in the document]
Comment (NehaKembalkarA10) [reply a10_octavia/controller/worker/flows/a10_load_balancer_flows.py]: Returning the vThunder model with the update acos_version in the "UpdateAcosVersionInVthunderEntry" task

**Review by ytsai-a10 — APPROVED** [review]

**Pull request merged**